Storage/src/Connection/Rest.php

```diff
@@ -626,6 +629,155 @@ private function sendWithRetry($resource, $method, $args)
             $args
         );
 
+        $requestHash = Uuid::uuid4()->toString();
+        $args['onRetryException'] = function (
+            \Exception $e,
+            $currentAttempt,
+            &$arguments
+        ) use ($requestHash) {
+            // Since we the the last attempt number here, so incrementing it
+            // to get the current attempt count
+            $this->updateRetryHeaders(
+                $arguments,
+                $requestHash,
+                $currentAttempt + 2
+            );
+        };
+
+        $args['onExecutionStart'] = function (&$arguments) use ($requestHash) {
+            $this->updateRetryHeaders(
+                $arguments,
+                $requestHash
+            );
+        };
+
```

Comment:
Instead of this here, I think we should extract this as a separate method like below as it would allow us to reuse this function resumable uploads, downloads as well as for sendWithRetry.

<details>

```php
// In insertObject()
$args = $this->addRetryHeaderCallbacks($args);
```

```php
// New method

/**
 * Adds callbacks to $args which amend retry hash and attempt to headers.
 * @param array $args
 *
 * @return array
 */
private function addRetryHeaderCallbacks(array $args): array
{
    $requestHash = Uuid::uuid4()->toString();
    $args['onRetryException'] = function (
        \Exception $e,
        $currentAttempt,
        &$arguments
    ) use ($requestHash) {
        // Since we the the last attempt number here, so incrementing it
        // to get the current attempt count
        $this->updateRetryHeaders(
            $arguments,
            $requestHash,
            $currentAttempt + 2
        );
    };
    $args['onExecutionStart'] = function (&$arguments) use ($requestHash) {
        $this->updateRetryHeaders(
            $arguments,
            $requestHash
        );
    };
    return $args;
}
```
</details>

WDYT @saranshdhingra ?

Reply:
Yeah we can do that. But let's do this once we merge both resumable downloads and the uploads PR.